lib/stages/ReadGain.cpp

```diff
@@ -0,0 +1,240 @@
+    if (_gain_dir_frb.length() > 0)
+        configUpdater::instance().subscribe(
+            config.get<std::string>(unique_name, "updatable_config/gain_frb"),
+            std::bind(&ReadGain::update_gains_frb_callback, this, _1));
+
+    // listen for gain updates PSR
+    configUpdater::instance().subscribe(
+        config.get<std::string>(unique_name, "updatable_config/gain_psr"),
+        std::bind(&ReadGain::update_gains_psr_callback, this, _1));
+}
+
+bool ReadGain::update_gains_frb_callback(nlohmann::json& json) {
+    {
+        std::lock_guard<std::mutex> lock(mux);
+        update_gains_frb = true;
+    }
+    cond_var.notify_all();
+
+    try {
+        _gain_dir_frb = json.at("frb_gain_dir");
+    } catch (std::exception& e) {
+        WARN("[FRB] Fail to read gain_dir {:s}", e.what());
+        return false;
+    }
+    INFO("[ReadGain] updated gain with {:s}============update_gains={:d}", _gain_dir_frb,
+         update_gains_frb);
+    return true;
+}
+
+bool ReadGain::update_gains_psr_callback(nlohmann::json& json) {
+    {
+        std::lock_guard<std::mutex> lock(mux);
+        update_gains_psr = true;
+    }
+    cond_var.notify_all();
+
+    try {
+        _gain_dir_psr = json.at("pulsar_gain_dir").get<std::vector<string>>();
+        INFO("[PSR] Updating gains from {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s}",
+             _gain_dir_psr[0].c_str(), _gain_dir_psr[1].c_str(), _gain_dir_psr[2].c_str(),
+             _gain_dir_psr[3].c_str(), _gain_dir_psr[4].c_str(), _gain_dir_psr[5].c_str(),
+             _gain_dir_psr[6].c_str(), _gain_dir_psr[7].c_str(), _gain_dir_psr[8].c_str(),
+             _gain_dir_psr[9].c_str());
+    } catch (std::exception const& e) {
+        WARN("[PSR] Fail to read gain_dir {:s}", e.what());
+        return false;
+    }
+    return true;
+}
+
+void ReadGain::main_thread() {
+
+    if (first_pass) {
+        first_pass = false;
+        uint8_t* frame =
+            wait_for_full_frame(metadata_buf, unique_name.c_str(), metadata_buffer_precondition_id);
+        if (frame == NULL)
+            goto end_loop;
+        stream_id_t stream_id = get_stream_id_t(metadata_buf, metadata_buffer_id);
+        freq_idx = bin_number_chime(&stream_id);
+        freq_MHz = freq_from_bin(freq_idx);
+        metadata_buffer_precondition_id =
+            (metadata_buffer_precondition_id + 1) % metadata_buf->num_frames;
+    }
+    mark_frame_empty(metadata_buf, unique_name.c_str(), metadata_buffer_id);
+    metadata_buffer_id = (metadata_buffer_id + 1) % metadata_buf->num_frames;
+    unregister_consumer(metadata_buf, unique_name.c_str());
+
+    while (!stop_thread) {
+        DEBUG("update_frb_gains={:d} update_psr_gains={:d} (0=false 1=true)", update_gains_frb,
+              update_gains_psr);
+
+        // FRB================================================
+        {
+            std::unique_lock<std::mutex> lock(mux);
+            while (!update_gains_frb) {
```

Comment:
The way this is structured is problematic.  Right now if the pulsar gains are updated, but you are waiting in this loop, then you'll won't update the pulsar gains until the FRB gains are updated to break out of this loop and go to the next waiting condition.    There are two solutions here: 1) you can break this into two `stages`, or 2) make the condition variable sleep on `update_gains_frb` and `update_gains_psr`, and then break the loop if either are true, then have the updates below only happen if `update_gains_frb` or `update_gains_psr` are true.  If you do 2) I would also suggest breaking the gain file read logic into two functions for FRB and Pulsar just to make it easier to follow the logic.    Also I think 2 is likely less work to implement.